go/store/types/chunkedstring.go

```diff
@@ -0,0 +1,332 @@
+// Copyright 2020 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+import (
+	"context"
+	"errors"
+	"math"
+)
+
+type ChunkedString struct {
+	valueImpl
+}
+
+var _ Value = ChunkedString{}
+var _ valueReadWriter = ChunkedString{}
+
+//TODO: Consolidate all chunk size literals into global constants (this should be the smallest constant)
+const maxChunkedStringChunk = 1<<27 - 1024
+
+// NewChunkedString returns a new ChunkedString.
+func NewChunkedString(ctx context.Context, vrw ValueReadWriter, s string) (ChunkedString, error) {
+	strVal := String(s)
+	var strChunks []String
+	for true {
+		if len(strVal) > maxChunkedStringChunk {
+			strChunks = append(strChunks, strVal[:maxChunkedStringChunk])
+			strVal = strVal[maxChunkedStringChunk:]
+		} else {
+			strChunks = append(strChunks, strVal)
+			break
+		}
+	}
+
+	w := newBinaryNomsWriter()
+	err := ChunkedStringKind.writeTo(&w, vrw.Format())
+	if err != nil {
+		return ChunkedString{}, err
+	}
+	w.writeCount(uint64(len(strVal)))
+	w.writeCount(uint64(len(strChunks)))
+
+	for _, val := range strChunks {
+		valRef, err := NewRef(val, vrw.Format())
+		if err != nil {
+			return ChunkedString{}, err
+		}
+		valRef.vrw = vrw
+		targetVal, err := valRef.TargetValue(ctx, vrw)
+		if err != nil {
+			return ChunkedString{}, err
+		}
+		if targetVal == nil {
+			_, err = vrw.WriteValue(ctx, val)
+			if err != nil {
+				return ChunkedString{}, err
+			}
+		}
```

Comment:
What happens if the target value is not nil?

Reply:
Then the `Ref` hash points to a `Value` that already exists, so we don't need to write anything new.